libcst/matchers/_matcher_base.py

```diff
@@ -76,6 +84,11 @@ def __and__(
     def __invert__(self: _BaseMatcherNodeSelfT) -> "_BaseMatcherNodeSelfT":
         return cast(_BaseMatcherNodeSelfT, InverseOf(self))
 
+    def __hash__(self) -> int:
+        # Allow matchers whose attributes are sequences or mappings to be
+        # used as a dictionary key.
+        return hash(repr(self))
```

Comment:
Where does this diff add using matchers as dictionary keys?

This hash definition is kinda smelly and violates a number of the guidelines in  https://docs.python.org/3/reference/datamodel.html#object.__hash__ (e.g. you should never define `__hash__` on a class that does not also define `__eq__`, you should not define `__hash__` for mutable objects). It's hard to evaluate the actual risk without seeing why this is needed.

Reply:
Matchers are used as keys in matcher decorators. Without this, you can't have a matcher decorator that includes metadata matching. Possibly a dubious design decision previously on my part, but one that is present at the time of implementing this nonetheless. I briefly considered adding an `__eq__` and all of these are *technically* immutable in the fact that they're not designed to be messed with once constructed, but I guess your point is valid. I can do a bit of extra work to expose read-only properties or convert some of these types to frozen dataclasses (doesn't matter either way) and then make sure there's an `__eq__` as well. For concrete types, they're all frozen dataclasses so they're "guaranteed" to be immutable, but the point still stands. Matchers *should* be immutable because they match on LibCST nodes, which *are* immutable. I can shore that up and close the gap.

Reply:
Yeah, I'm not gonna be a stickler on runtime immutability if the intended uses are immutable, but if it's not hard to make them frozen dataclasses that would certainly be ideal.

I do think it would be better to define `__eq__` and to use an explicit set of fields for both equality and hashing rather than the rather implicit and potentially fragile route of relying on `repr`.

Reply:
I guess I'm still slightly confused why this is related to metadata, and why it needs to be added on `BaseMatcherNode` in this diff but wasn't needed before, even though BaseMatcherNode would have been used in matcher decorators before this too?

Reply:
Will do, that answers my internal question to myself on whether I should add `__eq__` or not.

Reply:
You're right that its not related to metadata as much as the issue was uncovered by the current implementation of metadata using a mapping. The problem exists in current code that we can't use match decorators in combination with matchers constructed with lists. I can split this out in the final PR.

Reply:
I think with that, I should implement `__eq__` hack on the `BaseMatcherNode` which compares in a similar manner, since I am satisfied with the `__repr__` that dataclasses generates. Essentially, this is all to get around having to codegen a `__eq__` and `__hash__` *per* generated class.

Reply:
Would it be fairly easy and more explicit / less hacky to dynamically get the list of fields from dataclass internal metadata for a generic `__hash__` rather than relying on `repr`? And then convert lists to tuples at that point before hashing?

What makes me slightly nervous is that in general it's eminently possible for two distinct classes/instances to end up with the same `repr`. Not sure if that could ever happen within LibCST. It's just definitely more robust to actually hash the fields' values instead of their reprs.

Reply:
Offline conversation resolved that likely the best way to handle this is to make matchers default equality and hashing based on identity, not value, the same way LibCST nodes do it.